app/js/ConnectedUsersManager.js

```diff
@@ -22,7 +22,12 @@ module.exports = {
   updateUserPosition(project_id, client_id, user, cursorData, callback) {
     logger.log({ project_id, client_id }, 'marking user as joined or connected')
 
-    const multi = rclient.multi()
+    // It is safe to use a pipeline instead of a multi here:
+    //  - others commands can change the clientsInProject hash without
+    //     impacting our operation
+    //  - other commands can update the connectedUser fields/expiry without
+    //     impacting any consistency requirements
+    const multi = rclient.pipeline()
```

Comment:
> autopieline

Interesting find :+1: 

> While waiting for pipeline response from Redis, Node will still be able to process requests. All commands issued by request handler will be enqueued in a new automatically created pipeline. This pipeline will not be sent to the server yet.
> 
> As soon as a previous automatic pipeline has received all responses from the server, the new pipeline is immediately sent without waiting for the events loop iteration to finish.

I would jump ship right away if these pipelines were flushed per event loop cycle. It seems like it can accumulate a lot of commands in the NodeJS process if some command in the previous pipeline blocks for too long.

---

> I feel like it's preferable to have a transaction as it simplifies the reasoning about what can happen. Are we just trying to save the MULTI + EXEC ops here?

Yes. In the best-case we have one write per cursor change. We would wrap one operation in a MULTI+EXEC then, which seems pretty wasteful.